include/boost/geometry/algorithms/detail/is_valid/pointlike.hpp

```diff
@@ -42,14 +45,26 @@ struct is_valid<Point, point_tag>
 // (have identical coordinate values in X and Y)
 //
 // Reference: OGC 06-103r4 (6.1.5)
-template <typename MultiPoint>
-struct is_valid<MultiPoint, multi_point_tag>
+template <typename MultiPoint, bool AllowEmptyMultiGeometries>
+struct is_valid<MultiPoint, multi_point_tag, AllowEmptyMultiGeometries>
 {
-    static inline bool apply(MultiPoint const& /*multipoint*/)
+    template <typename VisitPolicy>
+    static inline bool apply(MultiPoint const& multipoint,
+                             VisitPolicy& visitor)
     {
-        // we allow empty multi-geometries, so an empty multipoint
-        // is considered valid
-        return true;
+        if (AllowEmptyMultiGeometries || boost::size(multipoint) > 0)
+        {
+            // we allow empty multi-geometries, so an empty multipoint
+            // is considered valid
+            visitor.template apply<no_failure>();
+            return true;
+        }
+        else
+        {
+            // we do not allow an empty multipoint
+            visitor.template apply<failure_few_points>();
+            return false;
```

Comment:
I have to admit I did not think of is as an advantage, but you are right.

Here is one specific advantage I can think of: right now we allow duplicate points, and this is done via a template parameter that is by default set to `true` (allow duplicate points). If we switch the default value to `false`, then different visitors can allow or disallow duplicate points, and this can be done at run-time instead of compile time.

So here is what I propose: change the interface of the visitor's `apply` methods to return the boolean success/failure value, which will then be used by `is_valid()` as the return value.
Also change the default value of the `AllowDuplicates` template parameter of `dispatch::is_valid` to `false`, and modify the two visitors we have now to accept duplicate points as valid. The drawback I see here is that `is_valid()` will now check for duplicate points, which means a small time overhead (personally, I am okay with this additional overhead).

What do you think?